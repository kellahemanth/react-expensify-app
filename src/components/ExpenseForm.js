import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now);
// console.log(now.format('MMM Do, YYYY'))

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    changeDescription = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }
    changeNote = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    changeAmount = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d{0,10}(\.\d{0,2})?$/)){
            this.setState(() => ({amount})); 
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
           this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => this.setState(() => ({calendarFocused: focused}));
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({
                error: 'Make sure Description and Amount are not empty!'
            }));
        }else{
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        autoFocus
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.changeDescription}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.changeAmount}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                        displayFormat='MMM Do, YYYY'
                    />
                    <textarea
                        placeholder="Add a note(optional)"
                        value={this.state.note}
                        onChange={this.changeNote}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    };
}