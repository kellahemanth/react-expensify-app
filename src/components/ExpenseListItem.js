import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => (
    <div>
        <h3><Link to={`edit/${props.expense.id}`}>{props.expense.description}</Link></h3>
        <p>Amount: {props.expense.amount}</p>
        <p>Created At: {props.expense.createdAt}</p>
    </div>
);

export default ExpenseListItem;