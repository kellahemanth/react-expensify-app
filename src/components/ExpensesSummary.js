import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import expensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expenseCount, expenseTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedTotal = '\u20B9'+ numeral(expenseTotal).format('0.00');
    return(
        <div>
            { 
                expenseCount > 0 && 
                    <h1>Viewing {expenseCount} {expenseWord} totalling {formattedTotal}</h1>
            }
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses,state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: expensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);