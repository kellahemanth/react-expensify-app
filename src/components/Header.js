import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink>&nbsp;&nbsp;
        <NavLink to="/create" activeClassName="is-active">Create expense</NavLink>&nbsp;&nbsp;
    </header>
);

export default Header;