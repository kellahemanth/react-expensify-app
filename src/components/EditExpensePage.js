import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense } from '../actions/expenses';
import { startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component{
    onSubmit = (updatedExpense) => {
        this.props.startEditExpense(this.props.expense.id,updatedExpense);
        this.props.history.push("/");
    };
    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense);
        this.props.history.push("/");
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense = {this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    };
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, updatedExpense) => dispatch(startEditExpense(id,updatedExpense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);