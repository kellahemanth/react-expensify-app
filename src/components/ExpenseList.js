import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import Expenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? 
            (
                <p>No Expenses</p>
            ) : 
            (
                props.expenses.map((expense) => {
                    return (
                        <ExpenseListItem expense={expense} 
                            key={expense.id}
                        />
                    );
                })
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: Expenses(state.expenses,state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);
