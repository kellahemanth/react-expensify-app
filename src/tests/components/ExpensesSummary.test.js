import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should show expenseCount and expenseTotal correctly for multiple expenses', () => {
    const expenseCount = 2;
    const expenseTotal = 10;
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expenseTotal={expenseTotal}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should show expenseCount and expenseTotal correctly for single expense', () => {
    const expenseCount = 1;
    const expenseTotal = 15.5;
    const wrapper = shallow(<ExpensesSummary expenseCount={expenseCount} expenseTotal={expenseTotal}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should hide expenseCount and expenseTotal', () => {
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
});