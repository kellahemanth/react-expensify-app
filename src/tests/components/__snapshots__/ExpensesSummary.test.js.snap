// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should hide expenseCount and expenseTotal 1`] = `<div />`;

exports[`should show expenseCount and expenseTotal correctly for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    ₹10.00
  </h1>
</div>
`;

exports[`should show expenseCount and expenseTotal correctly for single expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    ₹15.50
  </h1>
</div>
`;
